### PR TITLE
fixed conversion to sparse for non-numeric index

### LINKED ISSUE #11633: Pivot to SparseDataFrame: TypeError: ufunc 'isnan' not supported in sparse matrix conversion
I want to convert a DataFrame to SparseDataFrame before pivoting it (when it gets really sparse, see also [this discussion](http://stackoverflow.com/questions/31661604/efficiently-create-sparse-pivot-tables-in-pandas) ). I have a textual key, which I need to keep ("chr"):

    df = pd.DataFrame( list(zip([3,2,4,1,5,3,2],
                 ["chr1", "chr1", "chr1",  "chr1", "chr2", "chr2", "chr3"], 
                [100,100, 100, 200, 1,3,1],
                [True, True, True, False, True, False, True],
                [-1,0,1,3, 0,2,1])) ,
                columns = ["counts", "chr", "pos", "strand", "distance"])
    
    df.iloc[:,1:].dtypes
    Out[]: 
    chr         object
    pos          int64
    strand        bool
    distance     int64
    dtype: object

For this small table it works well with regular `DataFrame`:

    pd.pivot_table(df, index= [ "chr", "pos"], columns= ["strand","distance"], values= "counts").fillna(0)

         strand   False    True       
    distance     2  3    -1  0  1
    chr  pos                     
    chr1 100     0  0     3  2  4
         200     0  1     0  0  0
    chr2 1       0  0     0  5  0
         3       3  0     0  0  0
    chr3 1       0  0     0  0  2
But I need to do it on much larger matrices. So I tried to do following trick:

    dfpiv = pd.pivot_table(pd.SparseDataFrame(df), index= [ "chr", "pos"], columns= ["strand","distance"], values= "counts")

but I am getting:

    TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

Are there any plans to include a functionality option into `pivot` function for automatic conversion into SparseDataFrame?

### PR DESCRIPTION
**Pull request by DSLituiev**
this fixes  #11633

this fix breaks some sparse tests,
stuck on `test_combine_first` : `ValueError: total size of new array must be unchanged`

### CODE REVIEW

**Comment by jreback** [comment]
this looks reasonable. I think you may have to adjust some tests, and maybe some more fixing of the dtype propogation.

**Comment by DSLituiev** [comment]
I fixed one test to its original state (accounting for conversion of integer values to floats with `to_sparse`). Now `./test_sparse.py` runs OK.

**Comment by jreback** [comment]
can you rebase when you have a chance.

**Comment by DSLituiev** [comment]
rebased

**Comment by jreback** [comment]
need some tests for this.

**Comment by DSLituiev** [comment]
I added a test. 
One existing test fails now.

File: pandas/sparse/tests/test_sparse.py [hunk 1]
```diff
@@ -1720,6 +1720,17 @@ def test_nan_columnname(self):
         nan_colname_sparse = nan_colname.to_sparse()
         self.assertTrue(np.isnan(nan_colname_sparse.columns[0]))
 
+    def test_pivot(self):
+        df = pd.DataFrame( list(zip([3,2,4,1,5,3,2],
```
Comment (jreback) [inline pandas/sparse/tests/test_sparse.py]: add the issue number as a comment; don't use ``pd.``

**Comment by DSLituiev** [comment]
The current fix still does not completely solve the problem of sparse pivot #11633 (it uses sparse as input, but produces a dense output). One will need to look into `pivot_table` itself to see if one can make it spit out a sparse table.

**Comment by jreback** [comment]
@DSLituiev that's ok, let's get this bug fix in, then can work on that (which is a bit more complicated). *you* can always start constructing a fix based on THIS branch if you would like.

**Comment by sinhrks** [comment]
@DSLituiev can you update?

**Comment by jreback** [comment]
@DSLituiev can you update, lots of Sparse changes recently.

**Comment by DSLituiev** [comment]
I'll have a look

**Comment by jreback** [comment]
can you rebase/update

**Comment by jreback** [comment]
can you rebase / update, we've had lots of sparse changes.

**Comment by sinhrks** [comment]
As it is important for #667, I've prepared #13201.

**Comment by jreback** [comment]
superseded by #13021

**Pull request closed without merging**